### PR TITLE
Refactor JH integration and UI

### PR DESCRIPTION
**Pull request by ioandr**
This is a WIP PR, the first in a series of PRs to refactor how Kubeflow integrates with JupyterHub, how one can have multiple UIs for JupyterHub, and how we can extend the default UI and the associated KubeSpanwer logic to handle Persistent Volume Claims.

The main changes of this PR are:
* It decouples the configuration of JupyterHub into a new file `jupyterhub_config.py` so it is distinct from the definition of class KubeFormSpawner, in `spawner.py`.
* It externalizes the existing Spawner UI as a set of files which include an HTML form template (`template.html`), Javascript code (`script.js`) and stylesheet (`style.css`). Our implementation borrows heavily from @pdmack’s work available [at this commit](https://github.com/pdmack/kubeflow/commit/6771a03ee04ea31570dd7be01ba634d84537b672).
* Define the default UI as the set of these files, under `kubeflow/core/ui/default`. At this point, we also acknowledge that the UI has associated backend logic, so we move `spawner.py` under the same directory. We define the UI as the set of these files.
* Make the default UI configurable and enable whoever deploys kubeflow to configure it easily by modifying a single YAML file, `config.yaml` under the UI-specific directory.
* Improve the default UI significantly, which include: Font Awesome icons, use the fonts proposed by @pdmack in the aforementioned commit, switch to Bootstrap panels, add a form heading, footer, and short per-field descriptions, provide feedback when pressing the “Spawn” button, and other minor changes. See the attached screenshot on how this looks. [Please note the screenshot also includes PVC-related fields which are *not* part of this PR, but consider this a preview of a second, upcoming WIP PR extending the proposed refactored UI mechanism :) ]

jupyterhub_default_ui

Making these changes solves a few standing issues:

* The admin can very easily choose between multiple UIs [“skins”], by setting the ksonnet `ui` parameter. This was the issue that @pdmack had started working on and was also discussed in previous community meetings.
* It simplifies making custom modifications to the UI (form fields, colors, etc, and creating new UIs, by externalizing the UI definition as distinct files, removing the need to edit Python code directly.
The admin can easily configure the UI by modifying its associated `config.yaml`, in a declarative way. Examples include to make certain fields read-only by the user.
* The default UI and its configuration has become quite easier on the eyes, aligned with the changes proposed at: https://github.com/kubeflow/kubeflow/pull/1551
[Something must be off with GitHub currently, because this PR is shown as “Merged”, but the text reads “k8s-ci-robot merged 0 commits into kubeflow:master from unknown repository 11 days ago”, which doesn’t make any sense. Nothing is shown in the git tree as well. Please comment if you know anything more.]
* It paves the way for a new WIP PR which will extend the current default UI so it can handle arbitrary volume claims, per #34.

Breaking changes:
* Including a single `config.yaml` obsoletes various ksonnet parameters, including `registry`, `repoName`.
Extending the default UI to support arbitrary PVCs will also obsolete a couple of parameters, including `disks`, `notebookPVCMount`.

Known issues / next steps:
* We have not yet modified the tests to exclude setting these parameters, this WIP PR does not yet pass the tests, but it will be the first commit to follow.
* We have not yet verified that this PR passes the [very] recently added test for JupyterHub at https://github.com/kubeflow/kubeflow/commit/6687c762620951c77763afc50dd27b8447288212
* We have only tested this PR on a on-prem K8s deployment; we plan to test it on GCP and Minikube platforms.
* The UI respects parameters set as `readOnly` in `config.yaml`, but the Python code must also enforce them when handling a new request.
* We need a way to document the format of `config.yaml`. For now, please see the contents of the included `kubeflow/core/ui/default/config.yaml` and the code in `kubeflow/core/ui/default/spawner.py` on the supported YAML attributes. Where would be the best place to document this explicitly, as the supported format for `config.yaml`?

We will follow this PR with a second WIP one which will:
* Extend the logic of the default UI under `core/ui/default` so it handles arbitrary PVCs [This was actually the initial goal of #34! ]
* Add documentation / a short tutorial on how one can create a new UI as an extension of an existing one
* Add a new child UI as an example, which demonstrates using features of Arrikto’s Rok software, as demoed at the community meeting of Sep 12, 2018.

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kubeflow/1839)

### CODE REVIEW

**Comment by pdmack** [comment]
This looks **_really_** promising based on the screen caps. It'll dig in on the PR more the next few days.

**Comment by tlkh** [comment]
Hi @ioandr, I accidentally nuked my own PR #1551 with by trying to automate pulls. Then somehow it got flagged as merged. Apologies! 

I'm really excited about this, and I'll be taking a look at it as well!

**Comment by jlewi** [comment]
@ioandr If this is ready for review; can you remove the WIP form the title? Otherwise folks might not pay as much attention to it.

**Comment by jlewi** [comment]
It looks like the jsonnet tests failed

```
jsonnet eval /mnt/test-data-volume/kubeflow-presubmit-unittests-1839-a582468-4133-bac2/src/kubeflow/kubeflow/kubeflow/core/tests/jupyterhub_test.jsonnet --jpath /mnt/test-data-volume/kubeflow-presubmit-unittests-1839-a582468-4133-bac2/src/kubeflow/kubeflow --jpath /mnt/test-data-volume/kubeflow-presubmit-unittests-1839-a582468-4133-bac2/src/kubeflow/testing/workflows/lib/v1.7.0/ 
cwd=/mnt/test-data-volume/kubeflow-presubmit-unittests-1839-a582468-4133-bac2/src/kubeflow/kubeflow/kubeflow/core/tests
INFO|2018-10-29T23:48:48|util.py:56| Subprocess output:

INFO|2018-10-29T23:48:48|util.py:62| RUNTIME ERROR: field does not exist: ui
INFO|2018-10-29T23:48:48|util.py:62| /mnt/test-data-volume/kubeflow-presubmit-unittests-1839-a582468-4133-bac2/src/kubeflow/kubeflow/kubeflow/core/jupyterhub.libsonnet:27:25-34	thunk <a>
INFO|2018-10-29T23:48:48|util.py:62| /mnt/test-data-volume/kubeflow-presubmit-unittests-1839-a582468-4133-bac2/src/kubeflow/kubeflow/kubeflow/core/jupyterhub.libsonnet:27:25-47	function <anonymous>
INFO|2018-10-29T23:48:48|util.py:62| /mnt/test-data-volume/kubeflow-presubmit-unittests-1839-a582468-4133-bac2/src/kubeflow/kubeflow/kubeflow/core/jupyterhub.libsonnet:27:25-47	object <a>
INFO|2018-10-29T23:48:48|util.py:62| std.jsonnet:1215:42-46	thunk <a>
INFO|2018-10-29T23:48:48|util.py:62| std.jsonnet:1215:42-54	function <anonymous>
INFO|2018-10-29T23:48:48|util.py:62| std.jsonnet:1215:42-54	function <aux>
INFO|2018-10-29T23:48:48|util.py:62| std.jsonnet:1218:15-31	function <anonymous>
INFO|2018-10-29T23:48:48|util.py:62| std.jsonnet:1219:11-23	function <anonymous>
INFO|2018-10-29T23:48:48|util.py:62| /mnt/test-data-volume/kubeflow-presubmit-unittests-1839-a582468-4133-bac2/src/kubeflow/kubeflow/kubeflow/core/tests/jupyterhub_test.jsonnet:(25:1)-(38:2)
```

You should be able to rerun that locally by running the command

```
jsonnet eval /mnt/test-data-volume/kubeflow-presubmit-unittests-1839-a582468-4133-bac2/src/kubeflow/kubeflow/kubeflow/core/tests/jupyterhub_test.jsonnet --jpath /mnt/test-data-volume/kubeflow-presubmit-unittests-1839-a582468-4133-bac2/src/kubeflow/kubeflow --jpath /mnt/test-data-volume/kubeflow-presubmit-unittests-1839-a582468-4133-bac2/src/kubeflow/testing/workflows/lib/v1.7.0/ 
```
   * But changing the paths to the paths for your local filesystem

**Comment by jlewi** [comment]
/assign @pdmack 
/assign @kkasravi 

Do you want to take the lead on reviewing this PR?

**Comment by jlewi** [comment]
@pdmack @kkasravi thoughts on reviewing this PR?

**Comment by kkasravi** [comment]
@jlewi @pdmack I'll take the lead on this. I probably won't have comments until this afternoon.

File: kubeflow/core/jupyterhub.libsonnet [hunk 1]
```diff
@@ -4,16 +4,27 @@
   new(_env, _params):: {
     local params = _env + _params,
 
+    local defaultSpawnerData() = {
```
Comment (kkasravi) [inline kubeflow/core/jupyterhub.libsonnet]: no need to define a function. Just `defaultSpawnerData = ` would work

File: kubeflow/core/jupyterhub_config.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+# This allows local host path mounts to be read/writable
+env_uid = os.environ.get('NOTEBOOK_UID')
+if env_uid:
+    c.KubeSpawner.singleuser_uid = int(env_uid)
+env_gid = os.environ.get('NOTEBOOK_GID')
+if env_gid:
+    c.KubeSpawner.singleuser_fs_gid = int(env_gid)
+access_local_fs = os.environ.get('ACCESS_LOCAL_FS')
+if access_local_fs == 'true':
+    def modify_pod_hook(spawner, pod):
```
Comment (kkasravi) [inline kubeflow/core/jupyterhub_config.py]: Is it possible to add set more pod attributes directly the way we're doing here? For example volumes, volume_mounts.

Comment (vkoukis) [reply kubeflow/core/jupyterhub_config.py]: Kam, thanks for taking the time to review this!

I'm not sure I understand the workflow you have in mind here.
The original `kubeform_spawner.py` file had code which we have left intact, below, to compute values for `volumes`, `volume_mounts`, and this PR does not interfere with this piece of pre-existing functionality.

So, can you give an example of what you have in mind and how the administrator would be able to use it?

One way we have been thinking about allowing the administrator to set generic notebook pod attributes is:
1. the administrator could be able to modify an initial pod definition, e.g., in a file named `<uidir>/pod.yaml`, so they can specify any attribute they want in a generic way
2. we could expose this file inside the JupyterHub
3. the UI could start from this definition, and apply to it all decisions made by the user via the form before spawning the pod.

But this seems to require changes all the way to `make_pod()`  in KubeSpawner, see https://github.com/jupyterhub/kubespawner/blob/f49b1bd7060d0ca6fdf3342bb47491f33f95d79d/kubespawner/objects.py#L220, so we feel it was a much more disruptive change, and definitely outside the scope of an initial PR. We can always tackle this as a next step of course.

**Comment by pdmack** [comment]
Apologies, I've been slammed with other commitments. I'll try to look at this later today.

File: kubeflow/core/jupyterhub.libsonnet (hunk 1 above)
Comment (ioandr) [reply kubeflow/core/jupyterhub.libsonnet]: Kam, we have pushed the change you suggested in our PR. Please take a look.

**Comment by jlewi** [comment]
/lgtm
/hold so @kkasravi and @pdmack can continue to review.

Any thoughts about E2E testing?

File: kubeflow/core/jupyterhub_config.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
+import os
```
[the remark this answers is not in the document]
Comment (ioandr) [reply kubeflow/core/jupyterhub_config.py]: Hi Jeremy, we have just pushed an extra commit that adds the docstring you suggested.

**Pull request merged**